Add JVM metrics to metrics exporter

joneug opened the pull request:
This PR uses the [`simpleclient_hotspot`](https://github.com/prometheus/client_java) library to add JVM metric collectors (e.g. metrics about memory and CPU usage, threads, classloading, etc.). More details on the included JVM metrics collectors can be found [here](https://prometheus.github.io/client_java/).

### Your pull request should…

* [x] have a concise title
* [x] ~[close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists~
* [x] [be against the correct branch](https://docs.opencast.org/develop/developer/#development-process/#acceptance-criteria-for-patches-in-different-versions)
* [x] include migration scripts and documentation, if appropriate
* [x] pass automated tests
* [x] have a clean commit history
* [x] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

Commit:
Add JVM metrics to metrics exporter

This PR uses the simpleclient_hotspot library to add JVM metric collectors
(e.g. metrics about memory and CPU usage, threads, classloading, etc.).

Commit:
Update docs

lkiesow commented on docs/guides/admin/docs/modules/metrics.md:
```diff
@@ -8,18 +8,19 @@ tools like [Prometheus](https://prometheus.io). The endpoint is available at `/m
-There should usually be no need for monitoring several serves of a cluster. All will produce identical metrics.
+Opencast metrics are retrieved directly from the service registry and describe the current work load of the Opencast 
+cluster. There should usually be no need for monitoring Opencast metrics on several servers of a cluster. All will
+produce identical metrics. The standard JVM metrics, however, are specific to each server.
```
This is now kind of confusing (This does A except that it actually does B).
Let's reword the whole section to make it easier to understand.
What about:

```suggestion
Opencast related metrics describe the whole cluster and will be identical on all nodes
while JVM metrics are specific to each node.
```

lkiesow commented:
Hi @joneug,
thank you for contributing to Opencast.
I noticed that you have not yet filed an [Individual Contributor License Agreement](https://www.apereo.org/licensing/agreements/icla). Doing that (once) helps us to ensure that Opencast stays free for all.
Please let me know if you have any questions regarding the CLA.

lkiesow commented on docs/guides/admin/docs/modules/metrics.md:
````diff
@@ -51,6 +52,10 @@ requests_total 1.0
 opencast_asset_manager_events{organization="mh_default_org",} 1.0
 ```
 
+Additionally, standard JVM metrics are exported providing information about e.g. memory and CPU usage, threads,
+classloading, etc. The JVM metrics collectors are documented [here](https://prometheus.github.io/client_java/).
+A corresponding monitoring mixin with dashboards and alerting rules can be found
+[here](https://github.com/grafana/jsonnet-libs/tree/master/jvm-mixin).
````
It's always nice to see where you go instead of having a generic “click here” link:

```suggestion
Additionally, standard JVM metrics are exported providing information about e.g. memory and CPU usage, threads,
classloading, etc. For more information, take a look at the [Java client's JavaDoc](https://prometheus.github.io/client_java/).
A corresponding monitoring mixin with dashboards and alerting rules can be found at
[github.com/grafana/jsonnet-libs](https://github.com/grafana/jsonnet-libs/tree/master/jvm-mixin).
```

Though I don't think the JavaDocs are actually that helpful for end-users and it would be nice to link a metrics descriptions instead.

lkiesow reviewed (commented):
That patch works as expected.

---

While it makes sense to now have this on all node as you say, it's currently only deployed on admin nodes. Do we want to have it in all nodes?

```
% grep metrics assemblies/karaf-features/src/main/feature/feature.xml
    <bundle start-level="82">mvn:org.opencastproject/opencast-metrics-exporter/${project.version}</bundle>
    <bundle start-level="82">mvn:org.opencastproject/opencast-metrics-exporter/${project.version}</bundle>
    <bundle start-level="82">mvn:org.opencastproject/opencast-metrics-exporter/${project.version}</bundle>
```

---

Would you mind targeting `r/10.x` instead? I think that's a great addition and does not introduce any breaking changes. You could even target `r/9.x` but  since 10 is about to be released, I don't think we have to.

Commit:
Update docs

Apply suggestions from code review

Co-authored-by: Lars Kiesow <lkiesow@uos.de>

Commit:
Update karaf-features to include metrics-exporter in all distributions

Commit:
Update docs

joneug commented:
Thanks for your review, @lkiesow! 

> I noticed that you have not yet filed an [Individual Contributor License Agreement](https://www.apereo.org/licensing/agreements/icla).

I did that this morning.

> While it makes sense to now have this on all node as you say, it's currently only deployed on admin nodes. Do we want to have it in all nodes?

I changed the `feature.xml` so that the metrics-exporter is included in all distributions.

> Would you mind targeting r/10.x instead?

I rebased onto `r/10.x`.

> Though I don't think the JavaDocs are actually that helpful for end-users and it would be nice to link a metrics descriptions instead.

I did not find end-user documentation for the JVM metrics so I added a section in the docs.

joneug commented:
When rebasing onto `r/10.x` I still left the changes from PRs #2663 and #2674.

lkiesow commented:
> I did that this morning.

Apereo got it → That checkbox is checked. 
Thanks

Will try doing a review later today (and hopefully merge it)

lkiesow commented:
Okay, it's not that easy. This unfortunately completely breaks several node types (like workers):

```
org.apache.felix.resolver.reason.ReasonException: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=opencast-worker; type=karaf.feature; version="[10.0.0.SNAPSHOT,10.0.0.SNAPSHOT]"; filter:="(&(osgi.identity=opencast-worker)(type=karaf.feature)(version>=10.0.0.SNAPSHOT)(version<=10.0.0.SNAPSHOT))" [caused by: Unable to resolve opencast-worker/10.0.0.SNAPSHOT: missing requirement [opencast-worker/10.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-core; type=karaf.feature [caused by: Unable to resolve opencast-core/10.0.0.SNAPSHOT: missing requirement [opencast-core/10.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-metrics-exporter; type=osgi.bundle; version="[10.0.0.SNAPSHOT,10.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-metrics-exporter/10.0.0.SNAPSHOT: missing requirement [opencast-metrics-exporter/10.0.0.SNAPSHOT] osgi.service; filter:="(objectClass=org.opencastproject.assetmanager.api.AssetManager)"; effective:=active]]]
        at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
        at org.apache.felix.resolver.ResolverImpl.doResolve(ResolverImpl.java:392)
        at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:378)
        at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:332)
        at org.apache.karaf.features.internal.region.SubsystemResolver.resolve(SubsystemResolver.java:257)
        at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:393)
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1062)
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:998)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve opencast-worker/10.0.0.SNAPSHOT: missing requirement [opencast-worker/10.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-core; type=karaf.feature [caused by: Unable to resolve opencast-core/10.0.0.SNAPSHOT: missing requirement [opencast-core/10.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-metrics-exporter; type=osgi.bundle; version="[10.0.0.SNAPSHOT,10.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-metrics-exporter/10.0.0.SNAPSHOT: missing requirement [opencast-metrics-exporter/10.0.0.SNAPSHOT] osgi.service; filter:="(objectClass=org.opencastproject.assetmanager.api.AssetManager)"; effective:=active]]
        at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
        ... 12 more
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve opencast-core/10.0.0.SNAPSHOT: missing requirement [opencast-core/10.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-metrics-exporter; type=osgi.bundle; version="[10.0.0.SNAPSHOT,10.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-metrics-exporter/10.0.0.SNAPSHOT: missing requirement [opencast-metrics-exporter/10.0.0.SNAPSHOT] osgi.service; filter:="(objectClass=org.opencastproject.assetmanager.api.AssetManager)"; effective:=active]
        at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
        ... 13 more
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve opencast-metrics-exporter/10.0.0.SNAPSHOT: missing requirement [opencast-metrics-exporter/10.0.0.SNAPSHOT] osgi.service; filter:="(objectClass=org.opencastproject.assetmanager.api.AssetManager)"; effective:=active
        at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
        ... 14 more
```

The reason for that is the static OSGi binding to the asset manager which does not exist on those nodes.
Three possible ways to go forward:

- Leave it on admin-like nodes only for now, adjusting the documentation to not suggest that it will work on other nodes
- Turn the static into a dynamic binding and exclude the metrics which are not available.
- Run different implementations on admin-like nodes and other nodes

joneug commented:
I am working on turning that into a dynamic binding as you suggested. As such, I split the module `metrics-exporter` into the two modules `metrics-exporter-api` and `metrics-exporter-impl`. The API module exports a service interface that can be used by other modules to register new metric collectors. Finally, I refactored the code related to the asset manager out of the `metrics-exporter-impl` module so that it is only loaded on admin nodes.

So far this works great at compile time. However, I am still struggeling with OSGi missing requirements when running Opencast. I hope to resolve these soon...

lkiesow commented:
OSGi itself supports dynamic bindings. No need for large refactoring. If you want to, we can take a short look at that together? May be easier than writing a short OSGi manual ;-)

Commit:
Make reference to asset manager dynamic

lkiesow commented:
Yep, that's what I meant.
Is this ready for another review?

joneug commented:
Yes. I tested it with the develop and worker distributions and I think it works as expected.

lkiesow commented on modules/metrics-exporter/src/main/java/org/opencastproject/metrics/impl/MetricsExporter.java:
```diff
@@ -226,8 +232,17 @@ public void setOrganizationDirectoryService(OrganizationDirectoryService organiz
+      unbind = "unsetAssetManager"
+  )
   public void setAssetManager(AssetManager assetManager) {
     this.assetManager = assetManager;
```
On a worker, this now shows:
```
...
jvm_memory_pool_collection_init_bytes{pool="G1 Survivor Space",} 0.0
# HELP opencast_asset_manager_events Events in Asset Manager
# TYPE opencast_asset_manager_events gauge
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
...
```

I'm wondering if it would work to initialize the asset manager metric in here as well:
```suggestion
    this.assetManager = assetManager;
    eventsInAssetManager = Gauge.build()
            .name("opencast_asset_manager_events")
            .help("Events in Asset Manager")
            .labelNames("organization")
            .register();
```

plus unregistering it in `unsetAssetManager()` and not making the class var final.

joneug replied:
I changed the code as suggested. Now the asset manager metrics are no longer registered unnecessarily on non-admin nodes (tested with worker).

Commit:
Make asset manager metrics registration dynamic

Co-authored-by: Lars Kiesow <lkiesow@uos.de>

lkiesow reviewed (approved):
Will merge after the technical meeting if no one complains there
